src/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/HPack/Huffman.cs

```diff
@@ -308,119 +284,153 @@ public static (uint encoded, int bitLength) Encode(int data)
-        /// The number of bits in <paramref name="data"/> that may contain an encoded symbol.
-        /// This is not the exact number of bits that encode the symbol. Instead, it prevents
-        /// decoding the lower bits of <paramref name="data"/> if they don't contain any
-        /// encoded data.
-        /// </param>
-        /// <param name="decodedBits">The number of bits decoded from <paramref name="data"/>.</param>
-        /// <returns>The decoded symbol.</returns>
-        public static int Decode(uint data, int validBits, out int decodedBits)
+        private static short[,] BuildDecodingArray()
         {
-            // The code below implements the decoding logic for a canonical Huffman code.
-            //
-            // To decode a symbol, we scan the decoding table, which is sorted by ascending symbol bit length.
-            // For each bit length b, we determine the maximum b-bit encoded value, plus one (that is codeMax).
-            // This is done with the following logic:
-            //
-            // if we're at the first entry in the table,
-            //    codeMax = the # of symbols encoded in b bits
-            // else,
-            //    left-shift codeMax by the difference between b and the previous entry's bit length,
-            //    then increment codeMax by the # of symbols encoded in b bits
-            //
-            // Next, we look at the value v encoded in the highest b bits of data. If v is less than codeMax,
-            // those bits correspond to a Huffman encoded symbol. We find the corresponding decoded
-            // symbol in the list of values associated with bit length b in the decoding table by indexing it
-            // with codeMax - v.
-
-            int codeMax = 0;
+            var array = new short[15, 256];
 
-            for (int i = 0; i < s_decodingTable.Length && s_decodingTable[i].codeLength <= validBits; i++)
+            int nextAvailableSubIndex = 1;
+            // loop through each entry in the encoding table and create entries for it in our decoding array
+            // we're only going to 255. the EOS pattern of 256 (which will not fit into a byte) will be left out.
+            // trying to decode EOS should result in a decoding exception as per the RFC
+            // A Huffman-encoded string literal containing the EOS symbol MUST be treated as a decoding error.
+            // http://httpwg.org/specs/rfc7541.html#rfc.section.5.2
+            for (int i = 0; i < 256; i++)
             {
-                (int codeLength, int[] codes) = s_decodingTable[i];
+                var tableEntry = s_encodingTable[i];    // keep from having to do "s_encodingTable[i]" everywhere
+                int currentArrayIndex = 0;              // which array are we working with
 
-                if (i > 0)
+                // loop for however many bytes the value occupies
+                for (int j = 0; j <= Math.Ceiling(tableEntry.bitLength / 8.0); j++)
```

Comment:
Using floating point math (Math.Ceiling) here is really inefficient. This can done with integer math just fine.

Reply:
Change made